Port ClassDB tests to use doctest

Xrayez opened the pull request:
See #40659.

Existing test macros are replaced as aliases for doctest macros locally for now: #40950.

Methods do not return `TestResult` because that's handled by doctest.

The test suite is/should be structured in a way so that if prerequisites are not satisfied, the other checks should not be done: implemented using `SUBCASE` in doctest (subcases are like test leaves).

Problem that I stumbled upon is that testing `ClassDB` requires quite a lot of stuff to be initialized from the engine side... (mostly `register_types()` for complete coverage).

Commit:
Port ClassDB tests to use doctest

Extracted the most minimal core initialization functionality from
`setup()` and `setup2()` so that `ClassDB` could be tested properly
(input, audio, rendering, physics etc, are excluded).

Display and rendering servers/singletons are not initialized at all.

Due to the fact that most subsystems are disabled, fixed various crashes in the
process (in order):
- `AcceptDialog` OK/cancel swap behavior (used `DisplayServer` while
  `register_scene_types()`);
- `make_default_theme` which depends on `RenderingServer`;
- `XRServer` singleton access while calling `register_modules_types()`;
- hidden bug in a way joypads are cleaned up (MacOS and Linux only).

Removed manual `ClassDB` init/cleanup calls from `test_validate_testing.h`.

ClassDB tests:

Co-authored-by: Ignacio Etcheverry <ignalfonsore@gmail.com>

Xrayez commented:
Current output specifically from ClassDB tests:
```
PS D:\src\godot> .\bin\godot.windows.tools.64.4.exe --test --source-file="*test_class_db*"
Godot Engine v4.0.dev.Xrayez.977da0044 - https://godotengine.org
ERROR: Can't find layer: VK_LAYER_KHRONOS_validation
   at: VulkanContext::_check_layers (drivers\vulkan\vulkan_context.cpp:157)
ERROR: Can't find layer: VK_LAYER_LUNARG_standard_validation
   at: VulkanContext::_check_layers (drivers\vulkan\vulkan_context.cpp:157)
ERROR: Can't find layer: VK_LAYER_GOOGLE_threading
   at: VulkanContext::_check_layers (drivers\vulkan\vulkan_context.cpp:157)
WARNING: GENERAL - Message Id Number: 0 | Message Id Name: Loader Message
        ReadDataFilesInRegistry: Registry lookup failed to get layer manifest files.
        Objects - 1
                Object[0] - VK_OBJECT_TYPE_INSTANCE, Handle 2519364979488
     at: VulkanContext::_debug_messenger_callback (drivers\vulkan\vulkan_context.cpp:133)
 
[doctest] doctest version is "2.4.0"
[doctest] run with "--help" for options
WARNING: Signal name conflicts with method: 'BaseButton.pressed'.
     at: TestClassDB::add_exposed_classes (D:\src\godot\tests\test_class_db.h:694)
WARNING: Signal name conflicts with property: 'CanvasItem.hide'.
     at: TestClassDB::add_exposed_classes (D:\src\godot\tests\test_class_db.h:694)
WARNING: Signal name conflicts with property: 'Control.minimum_size_changed'.
     at: TestClassDB::add_exposed_classes (D:\src\godot\tests\test_class_db.h:694)
Ignoring class 'GDScriptEditorTranslationParserPlugin' because it's not exposed
Ignoring class 'GDScriptNativeClass' because it's not exposed
Ignoring class 'IP_Unix' because it's not exposed
WARNING: Signal name conflicts with property: 'Input.joy_connection_changed'.
     at: TestClassDB::add_exposed_classes (D:\src\godot\tests\test_class_db.h:694)
Ignoring class 'PhysicsDirectBodyState2DSW' because it's not exposed
Ignoring class 'PhysicsDirectBodyState3DSW' because it's not exposed
Ignoring class 'PhysicsServer2DSW' because it's not exposed
Ignoring class 'PhysicsServer3DSW' because it's not exposed
WARNING: Signal name conflicts with property: 'ProximityGroup3D.broadcast'.
     at: TestClassDB::add_exposed_classes (D:\src\godot\tests\test_class_db.h:694)
Ignoring class 'RootMotionView' because it's not enabled
WARNING: Signal name conflicts with method: 'TouchScreenButton.pressed'.
     at: TestClassDB::add_exposed_classes (D:\src\godot\tests\test_class_db.h:694)
===============================================================================
[doctest] test cases:      1 |      1 passed |      0 failed |     50 skipped
[doctest] assertions:  39347 |  39347 passed |      0 failed |
[doctest] Status: SUCCESS!
```

fire commented on tests/test_class_db.h:
```diff
@@ -31,12 +31,800 @@
+			return p_arg_type.name == p_context.names_cache.vector3_type ||
+				   p_arg_type.name == Variant::get_type_name(p_val.get_type());
+		default:
+			ERR_PRINT("Unexpected Variant type: " + itos(p_val.get_type()));
```
Use doc tests's INFO or FAIL?

Xrayez replied:
I think this could be moved outside the method call, and `CHECK_MESSAGE` accordingly. `FAIL` fails the entire test suite, but we're interested in checking more assertions...

Xrayez replied:
Or those individual equality checks could be replaced with `CHECK_MESSAGE`, so you know exactly what failed.

fire replied:
You messaged that you resolved the other issues.

Xrayez replied:
Yeah I have yet to resolve this. 😛

Xrayez replied:
Yeah I overthink this, `FAIL` is good because that would denote we have a `Variant` type which is not handled yet (new type or obsolete), so this would immediately notify testers that the checks above must be updated accordingly.

fire commented on tests/test_class_db.h:
```diff
@@ -31,12 +31,800 @@
+		}
+
+		if (!ClassDB::is_class_exposed(class_name)) {
+			OS::get_singleton()->print("Ignoring class '%s' because it's not exposed\n", String(class_name).utf8().get_data());
```
Use INFO?

Xrayez replied:
Seems like `INFO` doesn't work if one doesn't assert stuff, using `MESSAGE` works even without `--success` switch.

fire commented on tests/test_class_db.h:
```diff
@@ -31,12 +31,800 @@
+				// Actually, more methods like this may be added in the future, which could return
+				// something different. Let's put this check to notify us if that ever happens.
+				if (exposed_class.name != r_context.names_cache.object_class || String(method.name) != "free") {
+					WARN_PRINT("Notification: New unexpected virtual non-overridable method found."
```
Use INFO?

Xrayez replied:
I think this could be converted to doctest's `WARN`.

Xrayez replied:
`WARN` alone doesn't register those messages if you don't run with `--success` switch, but `WARN_FALSE_MESSAGE` works, which also allows to check the condition above.

fire commented on tests/test_class_db.h:
```diff
@@ -31,12 +31,800 @@
+				// However registering a signal with a conflicting name is still considered wrong.
+				// Unfortunately there are some existing cases that are yet to be fixed.
+				// Until those are fixed we will print a warning instead of failing the test.
+				WARN_PRINT("Signal name conflicts with " + String(method_conflict ? "method" : "property") +
```
Use INFO?

Xrayez replied:
Same as above.

Xrayez commented:
```
Run ./bin/godot.linuxbsd.opt.tools.64.mono --test
Godot Engine v4.0.dev.mono.custom_build.e0f362003 - https://godotengine.org
ERROR: X11 Display is not available
   at: DisplayServerX11 (platform/linuxbsd/display_server_x11.cpp:3364)
Error: Can't open display: 
ERROR: Unable to create DisplayServer, all display drivers failed.
   at: setup2 (main/main.cpp:1395)
```

All tests pass if you run this locally. Have to bypass display server initialization for CI at least, then.

Xrayez commented:
All tests pass!

Current output:
```
PS D:\src\godot> .\bin\godot.windows.tools.64.4.exe --test --source-file="*test_class_db*"
[doctest] doctest version is "2.4.0"
[doctest] run with "--help" for options
===============================================================================
D:\src\godot\tests\test_class_db.h(806):
TEST SUITE: [ClassDB] Tests
TEST CASE:  [ClassDB] Add exposed classes, builtin types, and global enums

D:\src\godot\tests\test_class_db.h(698): WARNING: WARN_FALSE( (method_conflict || exposed_class.find_method_by_name(signal.name)) ) is NOT correct!
  values: WARN_FALSE( true )
  logged: Signal name conflicts with method: 'BaseButton.pressed.

D:\src\godot\tests\test_class_db.h(698): WARNING: WARN_FALSE( (method_conflict || exposed_class.find_method_by_name(signal.name)) ) is NOT correct!
  values: WARN_FALSE( true )
  logged: Signal name conflicts with property: 'CanvasItem.hide.

D:\src\godot\tests\test_class_db.h(698): WARNING: WARN_FALSE( (method_conflict || exposed_class.find_method_by_name(signal.name)) ) is NOT correct!
  values: WARN_FALSE( true )
  logged: Signal name conflicts with property: 'Control.minimum_size_changed.

D:\src\godot\tests\test_class_db.h(466): MESSAGE: Ignoring class 'GDScriptEditorTranslationParserPlugin' because it's not exposed.

D:\src\godot\tests\test_class_db.h(466): MESSAGE: Ignoring class 'IP_Unix' because it's not exposed.

D:\src\godot\tests\test_class_db.h(698): WARNING: WARN_FALSE( (method_conflict || exposed_class.find_method_by_name(signal.name)) ) is NOT correct!
  values: WARN_FALSE( true )
  logged: Signal name conflicts with property: 'Input.joy_connection_changed.

D:\src\godot\tests\test_class_db.h(698): WARNING: WARN_FALSE( (method_conflict || exposed_class.find_method_by_name(signal.name)) ) is NOT correct!
  values: WARN_FALSE( true )
  logged: Signal name conflicts with property: 'ProximityGroup3D.broadcast.

D:\src\godot\tests\test_class_db.h(472): MESSAGE: Ignoring class 'RootMotionView' because it's not enabled.

D:\src\godot\tests\test_class_db.h(698): WARNING: WARN_FALSE( (method_conflict || exposed_class.find_method_by_name(signal.name)) ) is NOT correct!
  values: WARN_FALSE( true )
  logged: Signal name conflicts with method: 'TouchScreenButton.pressed.

===============================================================================
[doctest] test cases:      1 |      1 passed |      0 failed |     50 skipped
[doctest] assertions:  39302 |  39302 passed |      0 failed |
[doctest] Status: SUCCESS!
```

Those "warnings" could be in fact converted to actual assertions, but in order to not block the CI results, the test case can be marked as `doctest::may_fail()`.

Xrayez commented:
Things I've done to make it work currently:

1. Extracted (what I think) the most minimal core initialization functionality from `setup()` and `setup2()` so that `ClassDB` could be tested properly (input, audio, rendering, physics etc, are excluded).
2. Display and rendering servers/singletons are not initialized at all with this. That lead to various crashes like (in order):
   - `AcceptDialog` OK/cancel swap behavior (used `DisplayServer` while `register_scene_types()`);
   - `make_default_theme` which depends on `RenderingServer`.
   - `XRServer` singleton access while calling `register_modules_types()`;
   - hidden bug in a way joypads are cleaned up (MacOS and Linux only);

I haven't found a clean way to make it self-contained for `ClassDB` tests specifically, so before testing I just run `test_setup()` and `test_cleanup()` afterwards... But as you see, that seems to cover all previous tests without issues, and no crashes when compiling with sanitizer either: #40920.

I think we need to come up with different test context...

neikeq commented on tests/test_class_db.h:
```diff
@@ -31,12 +31,800 @@
+				   p_arg_type.name == Variant::get_type_name(p_val.get_type());
+		default:
+			String msg = "Unexpected Variant type: " + itos(p_val.get_type());
+			TEST_FAIL(msg.utf8().get_data());
```
I see you changed this from `ERR_PRINT` to `TEST_FAIL`. The reason this wasn't a `TEST_FAIL` is because it only adds additional information, the actual test check is done by the caller of `arg_default_value_is_assignable_to_type` because the caller provides a more informative error message for the failed test: `Invalid default value for parameter '{paramName}' of method '{className}.{methodName}'`. Will the `TEST_FAIL` here prevent that test fail message from being displayed?

Xrayez replied:
Yeah in fact it will fail the entire test case. Is there really a reason to continue if we stumble upon unknown `Variant` type in the first place?

But yeah I could just convert this to `MESSAGE` or `WARN_MESSAGE`, because doctest doesn't provide a way to log an *error* without failing, and `ERR_PRINT` suggests this fail IMO.

Also, `arg_default_value_is_assignable_to_type` is only used in `validate_method`.

Xrayez replied:
Nevermind me, I checked the context of where this is used, makes sense (is this for `DEFVAL`?). But I think the `Variant` type validation should be handled by the `Variant` class itself somewhere perhaps.

Xrayez replied:
Example output, should be good:

```
D:\src\godot\tests\test_class_db.h(406): ERROR: CHECK_FALSE( !arg_defval_assignable_to_type ) is NOT correct!
  values: CHECK_FALSE( true )
  logged: Invalid default value for parameter 'weight_scale' of method 'AStar.add_point'. Unexpected Variant type: 3
```

neikeq replied:
Looks good!

Xrayez commented:
Some discussion from IRC #godotengine-devel with @RevoluPowered:

```
Xrayez> RevoluPowered: since you're here, what do you think about initializing more stuff for all tests for now? Currently successfully ported ClassDB tests: #40980. I think we need to come up with different test context eventually (unit vs integration etc).
RevoluPowered> I don't want to, but totally get why you need to do it
RevoluPowered> if it's more practical for now
RevoluPowered> just do the init step basically
RevoluPowered> it's fine
RevoluPowered> later we can make it more decoupled together
RevoluPowered> like ideally we fix some of the problems in the engine, basically it assumes initialization of a lot of systems
RevoluPowered> I know the pain, I had some of it with OS singleton cleanup
RevoluPowered> I put a 'todo' :P there is a leak in tests on shutdown
Xrayez> Yeah I think I managed to clean it up in fact. :D
Xrayez> I mean, I think we just need to initialize a bit more than OS and StringName.
RevoluPowered> yeah definitely
RevoluPowered> I thought we could eventually compartmentalize it so if you need Audio, you don't need GDScript for instance
Xrayez> The problem is that doctest has no notion of setup/teardown, but that still can be implemented for each test case with SUBCASE and whatnot.
Xrayez> But yeah loading up required subsystems is a needed mechanism.
RevoluPowered> yeah i think he went with 'they can write init() shutdown()`
```

As long as we don't add more to the `test_setup()`, we should be fine for now IMO, because all tests pass currently. Testing stuff like `Input` obviously needs initialization of other systems and subsystems. If we reach that point, we'll need to come up with something better, I might write some proposal about this in fact.

Same for rendering, which is currently disabled for testing.

So, I'll mark this as ready for review now.

Xrayez commented:
Initially ported some GDScript tests: #41074, stumbled upon similar/same issues as in this PR. The GDScript implementation requires `ProjectSettings` to be initialized (setting up warnings etc). The problem with initializing this outside of `Main` is that `globals` is `static` (you'd have to provide some public methods to allow initialization and cleanup of those singletons and other objects).

Again see godotengine/godot-proposals#1307 to discuss how these problems could be practically resolved.

akien-mga commented on modules/mobile_vr/register_types.cpp:
```diff
@@ -37,7 +37,9 @@ void register_mobile_vr_types() {
 	mobile_vr.instance();
-	XRServer::get_singleton()->add_interface(mobile_vr);
+	if (XRServer::get_singleton()) {
+		XRServer::get_singleton()->add_interface(mobile_vr);
```
Do we risk leaking `mobile_vr` by not adding it as interface here? If so, maybe this should abort further up before instanciating it.

Xrayez replied:
I think not, but avoiding instantiation in the first place is a nice idea anyways, moved under condition. `ClassDB::register_class` is still needed for testing though.

It would be nice to merge #40920 to add those leak checks for that matter.

akien-mga commented:
CC @RevoluPowered 

It seems fine to me, is this good to merge for others?

fire commented:
I'm not able to look into detail, but it looks fine. The point of the tests is because the tested code is too complicated to think through. Also I'll try to post a pvs-studio report on Godot Engine master in a bit.

neikeq reviewed (commented):
The test code looks good. I can't comment on the test setup and additional changes made for it.

Xrayez commented:
Tests started to fail after rebase, might be some regressions in Godot (or new functionality), the crashes seem to be related to project settings, I'll check later.

Xrayez commented:
Regarding recent rebase, the crash above is a result of initializing `ClassDB` twice, I completely forgot about recently merged #40945 where I had to do that directly in the test case itself, so of course this is not needed anymore so I removed those init/cleanup calls from `test_validate_testing.h` in this PR.

But there's another visual problem which I misinterpreted as a silent crash (no output), namely when running tests through console on Windows: #41307, yet this doesn't block the PR.

Also rebased after #40920, so as you see, no memory leaks either.

> The point of the tests is because the tested code is too complicated to think through.

This PR opens up the door to testing all types which inherit from `Object`, and we have lots of those classes (godotengine/godot-docs#3153) to unit-test (far more than core data structures). IIRC @Calinou tried to port `Gradient` and we have identified the same crash problem that `ClassDB` has to be initialized first (or namely `ObjectDB`, which is also registered during `register_core_types()`, along with previously manually initialized stuff like `StringName`. In any case, I think it's not quite practical to attempt to decouple all of those subsystems currently (and perhaps they're not designed to be decoupled in the first place), and I'd like to avoid over-engineering this unless we do need to test more stuff like rendering/audio etc (if you ask me, that's unlikely to happen with Godot's rapid development process, unless someone is motivated enough to delve into this topic).

That said, all tests currently pass!

akien-mga commented:
Thanks!
